File: Source/LAffineTransform.cpp
/************************************************************

 LAffineTransform.cpp

    @author Christophe Berbizier (cberbizier@peersuasive.com)
    @license GPLv3
    @copyright 


(c) 2014, Peersuasive Technologies

*************************************************************/

// TODO: implement this in lua

#include "LAffineTransform_inh.h"

////// static methods
const char LAffineTransform::className[] = "LAffineTransform";
const Luna<LAffineTransform>::PropertyType LAffineTransform::properties[] = {
    { "identity", &LAffineTransform::getIdentity, &LBase::readOnly },
    {0,0}
};
const Luna<LAffineTransform>::FunctionType LAffineTransform::methods[] = {
    method( LAffineTransform, rotation ),
    method( LAffineTransform, translated ),
    method( LAffineTransform, getScaleFactor ),
    method( LAffineTransform, withAbsoluteTranslation ),
    method( LAffineTransform, isSingularity ),
    method( LAffineTransform, translation ),
    method( LAffineTransform, getTranslationX ),
    method( LAffineTransform, verticalFlip ),
    method( LAffineTransform, shear ),
    method( LAffineTransform, rotated ),
    method( LAffineTransform, scaled ),
    method( LAffineTransform, isOnlyTranslation ),
    method( LAffineTransform, followedBy ),
    method( LAffineTransform, getTranslationY ),
    method( LAffineTransform, fromTargetPoints ),
    method( LAffineTransform, sheared ),
    method( LAffineTransform, scale ),
    method( LAffineTransform, inverted ),
    method( LAffineTransform, isIdentity ),
    method( LAffineTransform, transformPoint ),
    method( LAffineTransform, transformPoints ),
    {0,0}
};

LAffineTransform::LAffineTransform(lua_State *L)
    : LBase(L, "LAffineTransform", true),
      AffineTransform()
{
    if ( lua_isstring(L, 2) )
        myName( lua_tostring(L, 2) );
}

LAffineTransform::LAffineTransform(lua_State *L, const AffineTransform& class_)
    : LBase(L, "LAffineTransform", true),
      AffineTransform( class_ )
{
    if ( lua_isstring(L, 2) )
        myName( lua_tostring(L, 2) );
}

LAffineTransform::~LAffineTransform() {}

int LAffineTransform::getIdentity(lua_State *L) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::identity
    ));
}

AffineTransform LAffineTransform::fromLuce( float mat00, float mat01, float mat02, 
        float mat10, float mat11, float mat12) 
{
    return { mat00, mat01, mat02, mat10, mat11, mat12 };
}

AffineTransform LAffineTransform::fromLuce( const Array<float>& a )
{
    return { a[0], a[1], a[2], a[3], a[4], a[5] };
}

int LAffineTransform::dump(lua_State*) {
    return LUA::returnTable( std::list<var>{ mat00, mat01, mat02, mat10, mat11, mat12 } );
}

/////// getters
int LAffineTransform::rotation ( lua_State *L ) {
    float angleInRadians = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::rotation( angleInRadians )
        ));
        
    float pivotX = LUA::getNumber<float>(2);
    float pivotY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(LUA::Get(),
        AffineTransform::rotation( angleInRadians, pivotX, pivotY )
    ));
}

int LAffineTransform::translated ( lua_State *L ) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::translated( LUA::getPoint<float>(2) )
    ));
}

int LAffineTransform::getScaleFactor ( lua_State* ) {
    return LUA::returnNumber( AffineTransform::getScaleFactor() );
}

int LAffineTransform::withAbsoluteTranslation ( lua_State* ) {
    float translationX = LUA::getNumber<float>(2);
    float translationY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(LUA::Get(),
        AffineTransform::withAbsoluteTranslation( translationX, translationY )
    ));
}

int LAffineTransform::isSingularity ( lua_State* ) {
    return LUA::returnBoolean( AffineTransform::isSingularity() );
}

int LAffineTransform::translation ( lua_State *L ) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::translation( LUA::getPoint<float>(2) )
    ));
}

int LAffineTransform::getTranslationX ( lua_State* ) {
    return LUA::returnNumber( AffineTransform::getTranslationX() );
}

int LAffineTransform::verticalFlip ( lua_State *L ) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::verticalFlip( LUA::getNumber<float>(2) )
    ));
}

int LAffineTransform::shear ( lua_State *L ) {
    float shearX = LUA::getNumber<float>(2);
    float shearY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::shear( shearX, shearY )
    ));
}

int LAffineTransform::rotated ( lua_State *L ) {
    float angleInRadians = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::rotated( angleInRadians )
        ));

    float pivotX = LUA::getNumber<float>(2);
    float pivotY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::rotated( angleInRadians, pivotX, pivotY )
    ));
}

int LAffineTransform::scaled ( lua_State *L ) {
    float factorX = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::scaled( factorX )
        ));

    float factorY = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L, 2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::scaled( factorX, factorY )
        ));

    float pivotX = LUA::getNumber<float>(2);
    float pivotY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::scaled( factorX, factorY, pivotX, pivotY )
    ));
}

int LAffineTransform::isOnlyTranslation ( lua_State* ) {
    return LUA::returnBoolean( AffineTransform::isOnlyTranslation() );
}

int LAffineTransform::followedBy ( lua_State *L ) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::followedBy( *LUA::from_luce<LAffineTransform>(2) )
    ));
}

int LAffineTransform::getTranslationY ( lua_State* ) {
    return LUA::returnNumber( AffineTransform::getTranslationY() );
}

int LAffineTransform::fromTargetPoints ( lua_State *L ) {
    float x00 = LUA::getNumber<float>(2);
    float y00 = LUA::getNumber<float>(2);
    float x10 = LUA::getNumber<float>(2);
    float y10 = LUA::getNumber<float>(2);
    float x01 = LUA::getNumber<float>(2);
    float y01 = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::fromTargetPoints( x00, y00, x10, y10, x01, y01 )
        ));

    float targetX2 = LUA::getNumber<float>(2);
    float targetY2 = LUA::getNumber<float>(2);
    float sourceX3 = LUA::getNumber<float>(2);
    float sourceY3 = LUA::getNumber<float>(2);
    float targetX3 = LUA::getNumber<float>(2);
    float targetY3 = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::fromTargetPoints( x00, y00, x10, y10, x01, y01, 
            targetX2, targetY2, sourceX3, sourceY3, targetX3, targetY3 )
    ));
}

int LAffineTransform::sheared ( lua_State *L ) {
    float shearX = LUA::getNumber<float>(2);
    float shearY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::sheared( shearX, shearY )
    ));
}

int LAffineTransform::scale ( lua_State *L ) {
    float factorX = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::scale( factorX )
        ));

    float factorY = LUA::getNumber<float>(2);
    if(lua_isnoneornil(L,2))
        return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
            AffineTransform::scale( factorX, factorY )
        ));

    float pivotX = LUA::getNumber<float>(2);
    float pivotY = LUA::getNumber<float>(2);
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::scale( factorX, factorY, pivotX, pivotY )
    ));
}

int LAffineTransform::inverted ( lua_State *L ) {
    return LUA::storeAndReturnUserdata<LAffineTransform>( new LAffineTransform(L,
        AffineTransform::inverted()
    ));
}

int LAffineTransform::isIdentity ( lua_State* ) {
    return LUA::returnBoolean( AffineTransform::isIdentity() );
}

/////// setters
int LAffineTransform::transformPoint ( lua_State* ) {
    // ValueType x = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y = *LUA::from_luce<LValueType>(2); // TODO;
    // AffineTransform::transformPoint( x, y );
    LUA::TODO_OBJECT( "transformPoint,  x, y " );
    lua_settop(LUA::Get(), 1); // added by TODO
    return 0;
}

int LAffineTransform::transformPoints ( lua_State* ) {
    // override
    // ValueType x1 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y1 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType x2 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y2 = *LUA::from_luce<LValueType>(2); // TODO;
    // AffineTransform::transformPoints( x1, y1, x2, y2 );

    // override
    // ValueType x1 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y1 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType x2 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y2 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType x3 = *LUA::from_luce<LValueType>(2); // TODO;
    // ValueType y3 = *LUA::from_luce<LValueType>(2); // TODO;
    // AffineTransform::transformPoints( x1, y1, x2, y2, x3, y3 );
    LUA::TODO_OBJECT( "transformPoints,  x1, y1, x2, y2, x3, y3 " );
    lua_settop(LUA::Get(), 1); // added by TODO
    return 0;
}

File: Source/lua/luce/LAffineTransform.lua
local luce, new = nil, nil
local myType = "LAffineTransform"
local mt = {}

local function hypot(a,b)
    math.sqrt(a*a + b*b)
end

local function limit(l, u, v)
    return (v < l) and l
        or ( (u < v) and upperLimit or v )
end
local function min(a, b)
    return (b < a) and b or a;
end

--[[ ************************************** ]]--

function mt:isIdentity()
    return  (self.mat01 == 0)
        and (self.mat02 == 0)
        and (self.mat10 == 0)
        and (self.mat12 == 0)
        and (self.mat00 == 1.0)
        and (self.mat11 == 1.0);
end

function mt:transformPoint(x, y)
    local x, y = x, y
    local oldX = x
    x = self.mat00 * oldX + self.mat01 * y + self.mat02
    y = self.mat10 * oldX + self.mat11 * y + self.mat12
    return x, y
end

function mt:transformPoints(x1, y1, x2, y2, x3, y3)
    local oldX1, oldX2, oldX3 = x1, x2, x3
    x1 = self.mat00 * oldX1 + self.mat01 * y1 + self.mat02
    y1 = self.mat10 * oldX1 + self.mat11 * y1 + self.mat12
    x2 = self.mat00 * oldX2 + self.mat01 * y2 + self.mat02
    y2 = self.mat10 * oldX2 + self.mat11 * y2 + self.mat12
    if(x3 and y3)then
        x3 = self.mat00 * oldX3 + self.mat01 * y3 + self.mat02
        y3 = self.mat10 * oldX3 + self.mat11 * y3 + self.mat12
    end
    return x1, y2, x2, y2, x3, y3
end

function mt:followedBy(other)
    return self:new{ other.mat00 * self.mat00 + other.mat01 * self.mat10,
                 other.mat00 * self.mat01 + other.mat01 * self.mat11,
                 other.mat00 * self.mat02 + other.mat01 * self.mat12 + other.mat02,
                 other.mat10 * self.mat00 + other.mat11 * self.mat10,
                 other.mat10 * self.mat01 + other.mat11 * self.mat11,
                 other.mat10 * self.mat02 + other.mat11 * self.mat12 + other.mat12
    }
end

function mt:translated(dx, dy)
    local dx, dy = dx, dy
    if not(dy)then dy = dx.y; dx = dx.x end
    return self:new{self.mat00, self.mat01, self.mat02 + dx,
                    self.mat10, self.mat11, self.mat12 + dy}

end

function mt:withAbsoluteTranslation(tx, ty)
    return self:new{ self.mat00, self.mat01, tx,
                     self.mat10, self.mat11, ty}

end

function mt:rotated(rad, pivotX, pivotY)
    if(pivotX and pivotY)then
        return self:followedBy( self:rotation(rad, pivotX, pivotY) )
    end
    local cosRad = math.cos (rad)
    local sinRad = math.sin (rad)
    return self:new{ cosRad * self.mat00 + -sinRad * self.mat10,
                     cosRad * self.mat01 + -sinRad * self.mat11,
                     cosRad * self.mat02 + -sinRad * self.mat12,
                     sinRad * self.mat00 +  cosRad * self.mat10,
                     sinRad * self.mat01 +  cosRad * self.mat11,
                     sinRad * self.mat02 +  cosRad * self.mat12}
end

function mt:scaled(factorX, factorY, pivotX, pivotY)
    if not(factorY)then
    return self:new{factorX * self.mat00, factorX * self.mat01, factorX * self.mat02,
                    factorX * self.mat10, factorX * self.mat11, factorX * self.mat12}

    elseif not(pivotX) then
    return self:new{factorX * self.mat00, factorX * self.mat01, factorX * self.mat02,
                    factorY * self.mat10, factorY * self.mat11, factorY * self.mat12}
    end
    return self:new{factorX * self.mat00, factorX 
                    * self.mat01, factorX * self.mat02 + pivotX * (1.0 - factorX),
                    factorY * self.mat10, factorY 
                    * self.mat11, factorY * self.mat12 + pivotY * (1.0 - factorY)}
end

function mt:sheared(shearX, shearY)
    return self:new{self.mat00 + shearX * self.mat10,
                    self.mat01 + shearX * self.mat11,
                    self.mat02 + shearX * self.mat12,
                    shearY * self.mat00 + self.mat10,
                    shearY * self.mat01 + self.mat11,
                    shearY * self.mat02 + self.mat12}
end

function mt:inverted()
    local determinant = (self.mat00 * self.mat11 - self.mat10 * self.mat01);
    if (determinant ~= 0) then
        determinant = 1.0 / determinant;
        local dst00 =  self.mat11 * determinant
        local dst10 = -self.mat10 * determinant
        local dst01 = -self.mat01 * determinant
        local dst11 =  self.mat00 * determinant

        return self:new{dst00, dst01, -self.mat02 * dst00 - self.mat12 * dst01,
                        dst10, dst11, -self.mat02 * dst10 - self.mat12 * dst11}
    else
        return self
    end
end

function mt:isSingularity()
    return (self.mat00 * self.mat11 - self.mat10 * self.mat01) == 0;
end

function mt:isOnlyTranslation()
    return  (self.mat01 == 0)
        and (self.mat10 == 0)
        and (self.mat00 == 1.0)
        and (self.mat11 == 1.0);
end

function mt:getTranslationX()
    return self.mat02
end

function mt:getTranslationY()
    return self.mat12
end

function mt:getScaleFactor()
    return (self.mat00 + self.mat11) / 2.0
end

-- [[ STATIC METHODS ]]--

local function translation(self, dx, dy)
    return self:new{1.0, 0, dx,
                    0, 1.0, dy}
end
mt.translation = translation

local function rotation(self, rad, pivotX, pivotY)
    if not(pivotX)then
    local cosRad = math.cos(rad)
    local sinRad = math.sin(rad)
    return self:new{cosRad, -sinRad, 0,
                      sinRad, cosRad, 0}
    else
    return self:new{cosRad, -sinRad, -cosRad * pivotX + sinRad * pivotY + pivotX,
                    sinRad, cosRad, -sinRad * pivotX + -cosRad * pivotY + pivotY}
    end
end
mt.rotation = rotation

local function scale(self, factorX, factorY, pivotX, pivotY)
    if not( factorY ) then
        return self:new{factorX, 0, 0, 0, factorX, 0}
    elseif not( pivotX ) then
        return self:new{factorX, 0, 0, 0, factorY, 0}
    end
    return self:new{factorX, 0, pivotX * (1.0 - factorX),
                    0, factorY, pivotY * (1.0 - factorY)}
end
mt.scale = scale

local function shear(self, shearX, shearY)
    return self:new{1.0, shearX, 0,
                    shearY, 1.0, 0}
end
mt.shear = shear

local function verticalFlip(height)
    return self:new{1.0, 0, 0, 0, -1.0, height}
end
mt.verticalFlip = verticalFlip

local function fromTargetPoints(...)
    if (#{...} == 6) then
    local x00, y00, x10, y10, x01, y01 = unpack{...}
    return self:new{x10 - x00, x01 - x00, x00,
                    y10 - y00, y01 - y00, y00}
    else
    local sx1, sy1, tx1, ty1,
          sx2, sy2, tx2, ty2,
          sx3, sy3, tx3, ty3 = unpack{...}
    return self:fromTargetPoints(sx1, sy1, sx2, sy2, sx3, sy3)
            :inverted()
            :followedBy (self:fromTargetPoints (tx1, ty1, tx2, ty2, tx3, ty3))
    end
end
mt.fromTargetPoints = fromTargetPoints

-- [[ INTERNAL ]] --

function mt:eq(self, other)
    return self.mat00 == other.mat00
        and self.mat01 == other.mat01
        and self.mat02 == other.mat02
        and self.mat10 == other.mat10
        and self.mat11 == other.mat11
        and self.mat12 == other.mat12
end

function mt:dump()
    return { self.mat00,
            self.mat01,
            self.mat02,
            self.mat10,
            self.mat11,
            self.mat12
    }
end

function mt:type()
    return self.__type
end

new = function(me, t, __type)
    local self = {}
    local t = t or {}
    self.mat00 = t[1] or t.mat00 or 1.0
    self.mat01 = t[2] or t.mat01 or 0
    self.mat02 = t[3] or t.mat02 or 0
    self.mat10 = t[4] or t.mat10 or 0
    self.mat11 = t[5] or t.mat11 or 1.0
    self.mat12 = t[6] or t.mat12 or 0
    self.__type = __type or t.__type or me.__type or "float"
    self.__ltype = myType
    return setmetatable(self, {
        __index = mt,
        __call = new,
        __self = myType,
        __tostring = function(self)
            return self.x..", "..self.y
        end,
        __eq = eq,
    })
end
mt.new = new

local __index = {
    identity = new{},
    translation = translation,
    rotation = rotation,
    scale = scale,
    shear = shear,
    verticalFlip = verticalFlip,
    fromTargetPoints = fromTargetPoints,
    new = new,
}
-- TODO: check if it works as expected !!!
setmetatable(mt, {__index=__index})

local xmeta = setmetatable({}, {
    __call = function(self,core,...)
        local self = self or {}
        luce = assert(core, "Missing luce core instance")
        self = setmetatable({}, {
            __call = new,
            __index = __index,
            __tostring = function()return myType end,
        })
        return self
    end
})

module(...)
return xmeta
